dspy/clients/lm.py

```diff
@@ -164,6 +165,55 @@ def copy(self, **kwargs):
         return new_instance
 
 
+@dataclass(frozen=True)
+class _ProviderAPIConfig:
+    """
+    API configurations for a provider (e.g. OpenAI, Azure OpenAI)
+    """
+
+    api_key: Optional[str]
+    api_base: Optional[str]
+    api_version: Optional[str]
+    # Azure OpenAI with Azure AD auth requires an Azure AD token for authentication.
+    # For all other providers, this field is empty
+    azure_ad_token: Optional[str]
+
+
+def _extract_provider_api_config(model: str, llm_kwargs: Dict[str, Any]) -> _ProviderAPIConfig:
+    """
+    Extract the API configurations from the specified LLM keyword arguments (`llm_kwargs`) for the
+    provider corresponding to the given model.
+
+    Note: The API configurations are removed from the specified `llm_kwargs`, if present, mutating
+    the input dictionary.
```

Comment:
This avoids duplicating code in `litellm_text_completion`, which pops the api key and base from the user-specified kwargs